extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/vault/AgroalVaultDynamicPassword.java

```diff
@@ -0,0 +1,77 @@
+package io.quarkus.agroal.runtime.vault;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Properties;
+import java.util.function.Supplier;
+
+import org.jboss.logging.Logger;
+
+import io.agroal.api.security.SimplePassword;
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.InstanceHandle;
+
+/**
+ * This a SimplePassword implementation that retrieves its password from vault 1.1 (hashicorp)
+ * using dynamically generated credentials
+ * The way to use it is simply by providing the following properties:
+ * <p>
+ * quarkus.vault.url=https://XXX:8200
+ * quarkus.vault.role=XXX
+ * quarkus.datasource.driver=XXX
+ * quarkus.datasource.url=jdbc:XXX
+ * quarkus.datasource.vault.dbrole=mydbrole
+ * <p>
+ * An instance of this class needs to be instantiated once per datasource.
+ * It provides a simple level of caching. In case of an error, it will return the last known value for the
+ * secret, and return again that value for the cache-period duration.
+ *
+ * @see VaultDataSourceRuntimeConfig
+ */
+public class AgroalVaultDynamicPassword extends SimplePassword {
+
+    private static final Logger log = Logger.getLogger(AgroalVaultDynamicPassword.class.getName());
+
+    private Supplier<Properties> supplier;
+
+    public AgroalVaultDynamicPassword(VaultDataSourceRuntimeConfig dsConfig) {
+        super("vault");
+        initSupplier(dsConfig.dbrole.get());
+    }
+
+    @Override
+    public Properties asProperties() {
+        return supplier.get();
+    }
+
+    // ---
+
+    private void initSupplier(String dbrole) {
+        try {
```

Comment:
@vsevel, why it is not injected ?

Reply:
I suppose you were thinking about moving the initSuplier() code into the producer?
done

Reply:
@vsevel, sorry, I meant why `VaultDbService` is accessed reflectively, given that there is a producer for it, the previous code for the DynamicPassword was fine otherwise

Reply:
because of the issues I encountered as discussed in https://github.com/quarkusio/quarkus/pull/2897#issuecomment-524668423
do you think that should have worked?